File: py/matrix.py
class matrix:
    def __init__(self, sz, mod, val=0):
        self.sz = sz
        self.mod = mod
        self.A = [[val % mod for _ in range(sz)] for _ in range(sz)]

    def __getitem__(self, i):
        return self.A[i]

    def size(self):
        return self.sz

    @staticmethod
    def identity(n, mod):
        I = matrix(n, mod)
        for i in range(n):
            I[i][i] = 1
        return I

    def __add__(self, other):
        assert self.sz == other.sz and self.mod == other.mod
        C = matrix(self.sz, self.mod)
        for i in range(self.sz):
            for j in range(self.sz):
                C[i][j] = (self[i][j] + other[i][j]) % self.mod
        return C

    def __iadd__(self, other):
        assert self.sz == other.sz and self.mod == other.mod
        for i in range(self.sz):
            for j in range(self.sz):
                self[i][j] = (self[i][j] + other[i][j]) % self.mod
        return self

    def __sub__(self, other):
        assert self.sz == other.sz and self.mod == other.mod
        C = matrix(self.sz, self.mod)
        for i in range(self.sz):
            for j in range(self.sz):
                C[i][j] = (self[i][j]-other[i][j])%self.mod
        return C

    def __isub__(self, other):
        assert self.sz == other.sz and self.mod == other.mod
        for i in range(self.sz):
            for j in range(self.sz):
                self[i][j] = (self[i][j]-other[i][j])%self.mod
        return self

    def __mul__(self, other):
        assert self.sz == other.sz and self.mod == other.mod
        C = matrix(self.sz, self.mod)
        for i in range(self.sz):
            for k in range(self.sz):
                if self[i][k] == 0:
                    continue
                aik = self[i][k]
                for j in range(self.sz):
                    C[i][j] = (C[i][j]+aik*other[k][j])%self.mod
        return C

    def __imul__(self, other):
        self.A = (self*other).A
        return self

    def pow(self, n):
        base = self
        R = matrix.identity(self.sz, self.mod)
        while n > 0:
            if n&1:
                R *= base
            base *= base
            n >>= 1
        return R

    def __repr__(self):
        return "\n".join(" ".join(map(str, row)) for row in self.A)

File: C++/matrix/matrix.cpp
template<typename T> class matrix{
private:
    using mat = vector<vector<T>>;
public:
    int sz;
    mat A;
    explicit matrix(int sz, T val = T()) : sz(sz), A(sz, vector<T>(sz, val)) {}

    vector<T>& operator [] (int i){
        return A[i];
    }

    const vector<T>& operator[] (int i) const{
        return A[i];
    }

    int size() const{
        return sz;
    }

    static matrix identity(int n){
        matrix I(n);
        for (int i = 0; i < n; i++) I[i][i] = T(1);
        return I;
    }

    friend istream& operator >> (istream& is, matrix& M){
        for(int i = 0; i < M.size(); i++){
            for(int j = 0; j < M.size(); j++){
                is >> M[i][j];
            }
        }
        return is;
    }

    matrix operator + (const matrix& B) const{
        assert(sz == B.sz);
        matrix C(sz);
        for (int i = 0; i < sz; i++){
            for (int j = 0; j < sz; j++){
                C[i][j] = A[i][j]+B[i][j];
            }
        }
        return C;
    }

    matrix& operator += (const matrix& B){
        assert(sz == B.sz);
        for (int i = 0; i < sz; i++){
            for (int j = 0; j < sz; j++){
                A[i][j] += B[i][j];
            }
        }
        return *this;
    }

    matrix operator - (const matrix& B) const{
        assert(sz == B.sz);
        matrix C(sz);
        for (int i = 0; i < sz; i++){
            for (int j = 0; j < sz; j++){
                C[i][j] = A[i][j]-B[i][j];
            }
        }
        return C;
    }

    matrix& operator -= (const matrix& B){
        assert(sz == B.sz);
        for (int i = 0; i < sz; i++){
            for (int j = 0; j < sz; j++){
                A[i][j] -= B[i][j];
            }
        }
        return *this;
    }

    matrix operator * (const matrix& B) const{
        assert(sz == B.sz);
        matrix C(sz, T());
        for (int i = 0; i < sz; i++){
            for (int k = 0; k < sz; k++){
                if (A[i][k] == T()) continue;
                for (int j = 0; j < sz; j++){
                    C[i][j] += A[i][k]*B[k][j];
                }
            }
        }
        return C;
    }

    matrix& operator *= (const matrix& B){
        assert(sz == B.sz);
        *this = (*this) *B;
        return *this;
    }

    matrix pow(long long n) const{
        matrix base = *this;
        matrix R = identity(sz);
        while (n > 0){
            if (n&1) R *= base;
            base *= base;
            n >>= 1;
        }
        return R;
    }
};